pkg/util/external/util.go

```diff
@@ -0,0 +1,116 @@
+	// with MachineSet or MachineDeployments.
+	TemplateSuffix = "Template"
+)
+
+// Get uses the client and reference to get an external, unstructured object.
+func Get(ctx context.Context, c client.Client, ref *corev1.ObjectReference, namespace string) (*unstructured.Unstructured, error) {
+	obj := new(unstructured.Unstructured)
+	obj.SetAPIVersion(ref.APIVersion)
+	obj.SetKind(ref.Kind)
+	obj.SetName(ref.Name)
+	key := client.ObjectKey{Name: obj.GetName(), Namespace: namespace}
+	if err := c.Get(ctx, key, obj); err != nil {
+		return nil, fmt.Errorf("failed to retrieve %s external object %q/%q: %w", obj.GetKind(), key.Namespace, key.Name, err)
+	}
+	return obj, nil
+}
+
+// GenerateTemplate input is everything needed to generate a new template.
+type GenerateTemplateInput struct {
+	// Template is the TemplateRef turned into an unstructured.
+	// +required
+	Template *unstructured.Unstructured
+
+	// TemplateRef is a reference to the template that needs to be cloned.
+	// +required
+	TemplateRef *corev1.ObjectReference
+
+	// Namespace is the Kubernetes namespace the cloned object should be created into.
+	// +required
+	Namespace string
+
+	// OwnerRef is an optional OwnerReference to attach to the cloned object.
+	// +optional
+	OwnerRef *metav1.OwnerReference
+
+	// Labels is an optional map of labels to be added to the object.
+	// +optional
+	Labels map[string]string
+}
+
+func GenerateTemplate(in *GenerateTemplateInput) (*unstructured.Unstructured, error) {
+	template, found, err := unstructured.NestedMap(in.Template.Object, "spec", "template")
+	if !found {
+		return nil, fmt.Errorf("missing Spec.Template on %v %q", in.Template.GroupVersionKind(), in.Template.GetName())
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to retrieve Spec.Template map on %v %q", in.Template.GroupVersionKind(), in.Template.GetName())
+	}
+
+	// Create the unstructured object from the template.
+	to := &unstructured.Unstructured{Object: template}
+	to.SetResourceVersion("")
+	to.SetFinalizers(nil)
+	to.SetUID("")
+	to.SetSelfLink("")
+	to.SetName(names.SimpleNameGenerator.GenerateName(in.Template.GetName() + "-"))
+	to.SetNamespace(in.Namespace)
+
+	if to.GetAnnotations() == nil {
+		to.SetAnnotations(map[string]string{})
+	}
+	annotations := to.GetAnnotations()
+	annotations[mapiv1.TemplateClonedFromNameAnnotation] = in.TemplateRef.Name
+	annotations[mapiv1.TemplateClonedFromGroupKindAnnotation] = in.TemplateRef.GroupVersionKind().GroupKind().String()
+	to.SetAnnotations(annotations)
+
+	// Set the owner reference.
+	if in.OwnerRef != nil {
+		to.SetOwnerReferences([]metav1.OwnerReference{*in.OwnerRef})
+	}
+
+	// Set the object APIVersion.
+	if to.GetAPIVersion() == "" {
+		to.SetAPIVersion(in.Template.GetAPIVersion())
+	}
+
+	// Set the object Kind and strip the word "Template" if it's a suffix.
+	if to.GetKind() == "" {
+		to.SetKind(strings.TrimSuffix(in.Template.GetKind(), TemplateSuffix))
+	}
+	return to, nil
```

Comment:
do the `Labels` from `GenerateTemplateInput` need to be applied in here somewhere?

Reply:
That's a very good point.
From what I can tell `Labels` are not relevant for either `GenerateTemplateInput` or for the actual `EMR` (which is created here), so I don't know if they should be copied or not.
IMO we should remove it altogether to avoid confusion.
@n1r1 what do you think,  Is there a purpose for the labels here that I'm missing  ?

Reply:
I don't see any immediate need for that but I can understand where it comes from and why someone might expect that.
I'm not sure what's the correct behavior here but for the sake of being consistent with upstream implementation (which doesn't add labels AFAIU), I would keep at as is.

If we think this should be changed we might want to do that upstream first.